lib/spack/spack/stage.py

```diff
@@ -45,18 +75,16 @@ def _first_accessible_path(paths):
                 if can_access(path):
                     return path
             else:
-                # The path doesn't exist so create it and adjust permissions
-                # and group as needed (e.g., shared ``$tempdir``).
-                prefix = os.path.sep
-                parts = path.strip(os.path.sep).split(os.path.sep)
-                for part in parts:
-                    prefix = os.path.join(prefix, part)
-                    if not os.path.exists(prefix):
-                        break
-                parent = os.path.dirname(prefix)
-                gid = os.stat(parent).st_gid
-                mkdirp(path, group=gid, default_perms='parents')
-
+                # Remove adjacent occurance of `$user` (e.g., `$tempdir/$user`
+                # where `$tempdir` includes `$user`).
+                user = getpass.getuser()
+                user_node = '{0}{1}'.format(os.path.sep, user)
+                double_user = '{0}{0}'.format(user_node)
+                if double_user in path:
+                    path = path.replace(double_user, user_node, 1)
+
+                # Now create the stage root with the proper group/perms.
+                _create_stage_root(path)
```

Comment:
I was under the impression that features/options for testing are undesirable.  Where might the check-only function be used outside the context of testing?

FWIW. There was a discussion, probably off-line, regarding the need for creating the root stage directory when this function is invoked.  Key issues, IIRC, are we no longer rely on `$spack/var/spack/stage` as a fallback and the continued need for a well-known source directory that must be available before staging (i.e., before staging directories are created).

Reply:
> I was under the impression that features/options for testing are undesirable

I think by this you mean functionality that is only activated during testing. I agree up to a point on that but my recommendation here is not a case of that: my recommendation is to decompose this logic into pieces that are more-easily tested.

> There was a discussion, probably off-line, regarding the need for creating the root stage directory when this function is invoked.

Perhaps the directory needs to exist at the point that this function would be called. If this function is replaced with two function calls (and perhaps those calls are aggregated into their own function) then that function can create the path at the desired time. In other words I don't think my request here conflicts with that constraint (although it would require a reinterpretation of the constraint if the function definitions change).

Reply:
> I think for testing purposes it would be ideal to be able to check the paths and return one without creating anything along the way. If it becomes problematic because it is challenging to suggest a path that works, I would suggest splitting into two functions:
> 
>     * one handles the business of examining each candidate path and stripping excess user entries (but otherwise returns all the paths)
> 
>     * one attempts to create each path and returns the first one that succeeds

@scheibelp, I created a `_resolve_paths` function that removes the `$user` entry if `$tempdir` already includes such a node during the process of canonicalizing the paths.  So I _think_ the solution I committed yesterday addresses your issue but would appreciate your feedback.